students/alex_whitty/session05/mailroom3.py

```diff
@@ -0,0 +1,165 @@
+    0: write_a_letter,
+    1: write_a_letter_all,
+}
+
+def dir_for_letter():
+
+    input("Hit <Enter/Return> key to save to the current working directory.")
+
+
+def thank_you_letter():
+
+    answer = input("Please enter Full Name.>>>")
+    if answer.lower() == 'list':
+        for donor in donor_db.keys():
+            print(donor)
+    else:
+        if answer in donor_db:
+            current_donor(answer)
+        else:
+            add_donor(answer)
+
+
+
+
+def letter_to_all_donors():
+    dir_for_letter()
+    for donor in donor_db.keys():
+        with open(f'{donor}.txt', 'wt') as letter:
+            #letter.write(write_a_letter_all(donor))
+            letter.write(switch_func_dict.get(1)(donor))
+
+
+
+def current_donor(donor):
+
+    amount = input("Please enter donor amount.>>>")
+    donor_db[donor].append(float(amount))
+    dir_for_letter()
+    with open(f'{donor}.txt', 'wt') as letter:
+        #letter.write(write_a_letter(donor, amount))
+        letter.write(switch_func_dict.get(0)(donor, amount))
+    print('Thanks, {} for your generous donation.'.format(donor))
+
+
+
+def sort_key(donor_db):
+    return donor_db[:]
+
+
+def add_donor(new_donor):
+
+
+    donor_amount = input("What is the donation amount?")
+    donor_db[new_donor] = [float(donor_amount)]
+    print(donor_db)
+    dir_for_letter()
+    with open(f'{new_donor}.txt', 'wt') as letter:
+        #letter.write(write_a_letter(new_donor, donor_amount))
+        letter.write(switch_func_dict.get(0)(new_donor, donor_amount))
+    print('Thanks, {} for your generous donation.'.format(new_donor))
+
+
+
+def report():
+    summary = []
+    headers = ["Donor Name", "Total Given", "Num Gifts", "Average Gift"]
+    print()
+    print("{:20} | {:10} | {:10} | {}".format(headers[0], headers[1], headers[2], headers[3]))
+    print("-" * 80)
+
+    for k, v in donor_db.items():
+        total = (sum(v))
+        times = (len(v))
+        avg = (sum(v) / len(v))
+        summary.append([k, total, times, avg])
+    summary.sort(key=lambda d: d[1], reverse=True)
+    for i in summary:
+        print("{:20}  ${:>12,.2f}{:>10}     ${:>14,.2f}".format(i[0], i[1], i[2], i[3]))
+        print("")
+
+
+def exit_program():
+    print("Good-Bye!")
+    sys.exit()  # exit the interactive script
+
+# Processing
+def main():
+    while True:
+        response = input(prompt)  # continuously collect user selection
+        # now redirect to feature functions based on the user selection
+        if response == "1":
+            thank_you_letter()
+        elif response == "2":
+            report()
+            pass
+        elif response == "3":
+            letter_to_all_donors()
+        elif response == "4":
+            exit_program()
+        else:
+            print("Not a valid option!")
+
+
+
+if __name__ == "__main__":
+    # don't forget this block to guard against your code running automatically if this module is imported
+    main()
+
+sorted(donor_db, key=sort_key)
```

Comment:
not sure what this is doing here.  It should be in the function that needs it or at the top of the file if you want it sorted on load.